### PR TITLE
Flesh out blockstore implementation

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Update the blockstore implementation based on improvements to implemented in base store

# Implementation

- support delete block functionality on blockstore
- add immutability option for stores where values cannot change
- add HashOnRead (though, I have no idea why one would use it)
- switch to Multihash impl

# For Discussion

Maybe we should implement AllKeysChan? It's a rarely used function (not anywhere in Lotus to my knowledge) that's a heavy lift.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -206,6 +208,11 @@ func (s *Store) Put(key []byte, value []byte) error {
 		// Two keys may point to same IndexKey (i.e. two CIDS same multihash),
 		// and they need to be treated as the same key.
 		if storedKey != nil {
+			// if we're not accepting updates, this is the point we bail --
+			// the identical key is in primary storage, we don't do update operations
+			if s.immutable {
+				return types.ErrKeyExists
+			}
```
Comment (gammazero) [inline store/store.go:215]: I think the index key is only the prefix. Now that you have retrieved `storedKey` (the full key), I think you need to compare `storedKey` to `key` to see if the full key stored under the `indexKey` is the same as `key` or is some other key that happens to have the same prefix.
```suggestion
			if s.immutable && bytes.Compare(key, storedKey) == 0 {
				return types.ErrKeyExists
			}
```

**Review by masih — APPROVED** [review]

File: store/store_test.go [hunk 2]
```diff
@@ -37,56 +37,101 @@ func initStore(t *testing.T, dir string) (*store.Store, error) {
...
+		flPath := filepath.Join(tempDir, "storethehash.index.free")
+		file, err := os.Open(flPath)
+		require.NoError(t, err)
+		t.Cleanup(func() { require.NoError(t, file.Close()) })
+
+		iter := freelist.NewFreeListIter(file)
+		// Check freelist -- no updates
+		_, err = iter.Next()
+		require.EqualError(t, err, io.EOF.Error())
+	})
```
Comment (gammazero) [inline store/store_test.go:129]: Include in this test two keys that have the same bit prefix to test my suggested changes.

File: storethehash.go [hunk 3]
```diff
@@ -123,7 +135,7 @@ func (bs *HashedBlockstore) GetSize(c cid.Cid) (int, error) {
 
 // Put puts a given block to the underlying datastore
 func (bs *HashedBlockstore) Put(blk blocks.Block) error {
-	err := bs.store.Put(blk.Cid().Bytes(), blk.RawData())
+	err := bs.store.Put(blk.Cid().Hash(), blk.RawData())
```
Comment (gammazero) [inline storethehash.go:138]: Be careful that the first few bits of the multihash are pure hash data, and not multihash algo identifier. Otherwise all multihashes, using the same type of hash will have the same prefix bits and will tank performance.  In the indexer core, this was a painful discovery (which you helped us understand), and the fix was [this](https://github.com/filecoin-project/go-indexer-core/blob/main/store/storethehash/storethehash.go#L499-L501).

**Review by gammazero — CHANGES_REQUESTED** [review]
I think there is at least one change needed regarding key comparison.

File: store/store.go (hunk 1 above)
Comment (hannahhoward) [reply store/store.go:215]: as I read the code for `getPrimaryKeyData`, storedKey != nil indicates it has already done this check and verified equality. I will admit I find the code for `getPrimaryKeyData` a bit hard to parse.

File: storethehash.go (hunk 3 above)
Comment (hannahhoward) [reply storethehash.go:138]: yea. this is actual block data, so CIDs should be pure, with no unexpected prefixes, and the multihash code decodes and uses only the digests.

File: store/store_test.go (hunk 2 above)
Comment (hannahhoward) [reply store/store_test.go:129]: will do to test my hypothesis about getPrimaryKeyData

File: store/store.go (hunk 1 above)
Comment (gammazero) [reply store/store.go:215]: Yes, confirmed:
https://github.com/ipld/go-storethehash/blob/master/store/store.go#L335-L337
